src/cascadia/TerminalControl/ControlInteractivity.cpp

```diff
@@ -257,15 +258,27 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
         else if (WI_IsFlagSet(buttonState, MouseButtonState::IsRightButtonDown))
         {
-            // CopyOnSelect right click always pastes
-            if (_core->CopyOnSelect() || !_core->HasSelection())
+            if (_core->CopyOnSelect())
             {
+                // CopyOnSelect:
+                // 1. keyboard selection? --> copy the new content first
+                // 2. right click always pastes!
+                if (_core->IsInQuickEditMode())
+                {
+                    CopySelectionToClipboard(shiftEnabled, nullptr);

```

Comment:
If we're `CopyOnSelect() == true` wouldn't we have already copied the selection at this point in `ControlInteractivity::PointerReleased`?

---

When is `IsInQuickEditMode() == false`? Whenever we do a keyboard selection? Would `IsInKeyboardSelectionMode` not be a better (more descriptive and modern) function name then?

Reply:
> If we're `CopyOnSelect() == true` wouldn't we have already copied the selection at this point in `ControlInteractivity::PointerReleased`?

Yes and no.
Yes, because that covers the scenario of (1) create a mouse selection then (2) releasing the mouse button. The content immediately gets copied to your clipboard. When you right-click, you paste that content, so this code isn't called at all because we're not in "quick edit mode".

No, because "quick edit mode" is specifically when you add these steps to the scenario above... (3) shift+right to extend the selection a bit. Then, when you right-click, we already copied the wrong contents! We copied the data when you let go, not when you're right-clicking. So this extra code just detects when you've done step 3 and we need to copy the contents again.

> When is `IsInQuickEditMode() == false`? Whenever we do a keyboard selection? Would `IsInMouseSelectionMode` (or inversely `IsInKeyboardSelectionMode`) not be a better (more descriptive and modern) function name then?

`QuickEditMode == true` --> you made a selection with the mouse, then modified it with the keyboard
`MarkMode == true` --> you toggled mark mode manually (created a selection at the cursor)

They're mutually exclusive. After a quick chat with Dustin, here's a new approach...
```
enum class SelectionMode
{
   Mouse = 0,
   Keyboard,
   Mark
};
```
We just have 3 modes in increasing order. All mutually exclusive, but all useful to know when to show the selection markers and what trickery could occur. That also lets me combine `IsInMarkMode` and `IsInQuickEditMode`. AND it also gets rid of the historical term of "quick edit mode".

Reply:
this approach only applies to 1.15+